## Quantifying chimeric reads in shRNA integration windows	

Reads reported in `_p1.Chimeric.out.junction` are filtered for those involving alignment to `shRNA` and then are further filtered to those falling within the called *integration windows*.

```bash
 % head test1_p1.Chimeric.out.junction.shRNA.filtered_to_integration_windows.tsv
chr6|33272627|-|NB501223:102:HMC2KBGXF:1:12106:10408:19739	chr6	33272627	-	chr_shRNA	4554	-	2	0	1	NB501223:102:HMC2KBGXF:1:12106:10408:19739	33272628	52S98M3301p40M109N92M123N19M	4502	52M98S	1	301	250	297	297	0
chr7|44881390|-|NB501223:102:HMC2KBGXF:1:12108:17982:4123	chr7	44881390	-	chr_shRNA	4554	-	2	2	1	NB501223:102:HMC2KBGXF:1:12108:17982:4123	44881391	49S31M6p119M	4505	49M31S	1	199	149	196	196	0
chr14|20343258|+|NB501223:102:HMC2KBGXF:2:12104:1833:19161	chr_shRNA	7341	-	chr14	20343258	+	0	0	0	NB501223:102:HMC2KBGXF:2:12104:1833:19161	7342	84S66M116p150M1S	20343259	66S84M	1	301	214	293	293	0
chr5|73643317|+|NB501223:102:HMC2KBGXF:2:12105:12463:13434	chr5	73643317	+	chr_shRNA	636	-	0	0	1	NB501223:102:HMC2KBGXF:2:12105:12463:13434	73643273	44M107S	488	148M-107p107M44S	1	301	258	295	295	0
chr6|33272627|-|NB501223:102:HMC2KBGXF:2:22104:1615:5946	chr6	33272627	-	chr_shRNA	4554	-	2	0	1	NB501223:102:HMC2KBGXF:2:22104:1615:5946	33272628	51S98M-68p69M3070N82M	4503	51M98S	1	300	252	298	298	0
chr7|44881390|-|NB501223:102:HMC2KBGXF:2:22105:25297:9464	chr7	44881390	-	chr_shRNA	4554	-	2	2	1	NB501223:102:HMC2KBGXF:2:22105:25297:9464	44881391	50S31M6p117M	4504	50M31S	1	198	147	195	195	0
chr6|33272627|-|NB501223:102:HMC2KBGXF:3:21605:4862:8320	chr6	33272627	-	chr_shRNA	4554	-	2	0	1	NB501223:102:HMC2KBGXF:3:21605:4862:8320	33272628	38S99M3070N13M7p67M81N83M	4516	38M112S	1	300	267	300	300	0
chr7|44881427|-|NB501223:102:HMC2KBGXF:4:11501:11670:18517	chr7	44881427	-	chr_shRNA	8003	-	-1	0	0	NB501223:102:HMC2KBGXF:4:11501:11670:18517	44881428	118M	7854	149M	1	267	147	263	263	0
chr7|44881390|-|NB501223:102:HMC2KBGXF:4:21410:21394:8282	chr7	44881390	-	chr_shRNA	4554	-	2	2	1	NB501223:102:HMC2KBGXF:4:21410:21394:8282	44881391	31S31M6p117M	4523	31M31S	2	179	147	176	176	0
chr1|71076794|+|NB501223:102:HMC2KBGXF:1:12110:4807:16670	chr_shRNA	7991	+	chr1	71076794	+	2	2	2	NB501223:102:HMC2KBGXF:1:12110:4807:16670	7758	151M47p35M116S	71076795	35S83M1579N33M	1	302	186	299	299	0
```

These are further annotated with the *integration window*

```bash
 % head *windows.annotated.tsv
chr6|33272627|-|NB501223:102:HMC2KBGXF:1:12106:10408:19739	chr6:33272527-33272726|ENSG00000231500.7|RPS18|+	chr6	33272627	-	chr_shRNA	4554	-	2	0	1	NB501223:102:HMC2KBGXF:1:12106:10408:19739	33272628	52S98M3301p40M109N92M123N19M	4502	52M98S	1	301	250	297	297	0
chr7|44881390|-|NB501223:102:HMC2KBGXF:1:12108:17982:4123	chr7:44881290-44881489|ENSG00000146676.10|PURB|-	chr7	44881390	-	chr_shRNA	4554	-	2	2	1	NB501223:102:HMC2KBGXF:1:12108:17982:4123	44881391	49S31M6p119M	4505	49M31S	1	199	149	196	196	0
chr14|20343258|+|NB501223:102:HMC2KBGXF:2:12104:1833:19161	chr14:20343158-20343486|ENSG00000259001.3|RPPH1|-	chr_shRNA	7341	-	chr14	20343258	+	0	0	0	NB501223:102:HMC2KBGXF:2:12104:1833:19161	7342	84S66M116p150M1S	20343259	66S84M	1	301	214	293	293	0
chr5|73643317|+|NB501223:102:HMC2KBGXF:2:12105:12463:13434	chr5:73643062-73643416|ENSG00000214944.10|ARHGEF28|+	chr5	73643317	+	chr_shRNA	636	-	0	0	1	NB501223:102:HMC2KBGXF:2:12105:12463:13434	73643273	44M107S	488	148M-107p107M44S	1	301	258	295	295	0
chr6|33272627|-|NB501223:102:HMC2KBGXF:2:22104:1615:5946	chr6:33272527-33272726|ENSG00000231500.7|RPS18|+	chr6	33272627	-	chr_shRNA	4554	-	2	0	1	NB501223:102:HMC2KBGXF:2:22104:1615:5946	33272628	51S98M-68p69M3070N82M	4503	51M98S	1	300	252	298	298	0
chr7|44881390|-|NB501223:102:HMC2KBGXF:2:22105:25297:9464	chr7:44881290-44881489|ENSG00000146676.10|PURB|-	chr7	44881390	-	chr_shRNA	4554	-	2	2	1	NB501223:102:HMC2KBGXF:2:22105:25297:9464	44881391	50S31M6p117M	4504	50M31S	1	198	147	195	195	0
chr6|33272627|-|NB501223:102:HMC2KBGXF:3:21605:4862:8320	chr6:33272527-33272726|ENSG00000231500.7|RPS18|+	chr6	33272627	-	chr_shRNA	4554	-	2	0	1	NB501223:102:HMC2KBGXF:3:21605:4862:8320	33272628	38S99M3070N13M7p67M81N83M	4516	38M112S	1	300	267	300	300	0
chr7|44881427|-|NB501223:102:HMC2KBGXF:4:11501:11670:18517	chr7:44881290-44881489|ENSG00000146676.10|PURB|-	chr7	44881427	-	chr_shRNA	8003	-	-1	0	0	NB501223:102:HMC2KBGXF:4:11501:11670:18517	44881428	118M	7854	149M	1	267	147	263	263	0
chr7|44881390|-|NB501223:102:HMC2KBGXF:4:21410:21394:8282	chr7:44881290-44881489|ENSG00000146676.10|PURB|-	chr7	44881390	-	chr_shRNA	4554	-	2	2	1	NB501223:102:HMC2KBGXF:4:21410:21394:8282	44881391	31S31M6p117M	4523	31M31S	2	179	147	176	176	0
chr1|71076794|+|NB501223:102:HMC2KBGXF:1:12110:4807:16670	chr1:71076694-71076893|ENSG00000132485.14|ZRANB2|-	chr_shRNA	7991	+	chr1	71076794	+	2	2	2	NB501223:102:HMC2KBGXF:1:12110:4807:16670	7758	151M47p35M116S	71076795	35S83M1579N33M	1	302	186	299	299	0
```

Here the second column denotes the *integration window*, which also includes the gene at the *integration window* along with its strand.

`get_detailed_chimera_counts.py`python script is used to get per-*integration window* counts for all possible alignment combinations:

| host chromosome alignment strand | shRNA alignment strand |
| :------------------------------: | :--------------------: |
|                +                 |           +            |
|                +                 |           -            |
|                -                 |           +            |
|                -                 |           -            |

If there are 5 or more reads aligning to the host chromosome on the same strand as the gene annotation of the *integration window*, then this is considered as evidence for opposite strand transcription (OST). This is also reported in the counts table as follows:

```bash
 % head test5/test5_shRNA_integration_window_counts.tsv
integration_window	gene_strand	counts_++	counts_+-	counts_-+	counts_--	same_strand_counts	opposite_strand_counts	total_counts	opposite_strand_transcription
chr5:73643062-73643416|ENSG00000214944.10|ARHGEF28|+	+	0	678	1	0	678	1	679	True
chr5:113546853-113547052|ENSG00000047188.16|YTHDC2|+	+	0	130	0	0	130	0	130	True
chr5:73642533-73642853|ENSG00000214944.10|ARHGEF28|+	+	0	2455	0	0	2455	0	2455	True
chr6:52995534-52996006|ENSG00000202198.1|7SK|+	+	0	1	5	1146	1	1151	1152	False
chr14:20343158-20343486|ENSG00000259001.3|RPPH1|-	-	1	37	0	0	0	38	38	False
chrM:1946-2145|ENSG00000210082.2|MT-RNR2|+	+	0	1	0	36	1	36	37	False
chr13:95149166-95149365|ENSG00000125257.16|ABCC4|-	-	0	0	0	11	11	0	11	True
chr14:49862461-49862918|ENSG00000282885.2|RP11-596C23.6|-	-	2	657	0	0	0	659	659	False
chr14:49586487-49586838|ENSG00000276168.1|RN7SL1|+	+	0	0	1	530	0	531	531	False
```

Using `create_counts_matrix.v2.py` the total counts for each *integration window* can be aggregrated across sample to create a count matrix

```bash
 % head integration_windows.counts_matrix.v2.tsv                                   
integration_window	gene	nearest_verified_site	test1	test2	test3	test4	test5	test6	test7	test8
chr1:632158-632357	ENSG00000237973.1|MTCO1P12|+	chr1:71076570-71076570|+|ENSG00000132485.14|ZRANB2|70444213	0.0	0.0	0.0	0.0	4	7	1	4.0
chr1:28507280-28507479	ENSG00000274266.1|SNORA73A|+	chr1:71076570-71076570|+|ENSG00000132485.14|ZRANB2|42569091	0.0	0.0	0.0	1.0	9	13	12	14.0
chr1:71072449-71072648	ENSG00000132485.14|ZRANB2|-	chr1:71076570-71076570|+|ENSG00000132485.14|ZRANB2|3922	0.0	0.0	0.0	2.0	7	9	8	2.0
chr1:71076470-71076669	ENSG00000132485.14|ZRANB2|-	chr1:71076570-71076570|+|ENSG00000132485.14|ZRANB2|0	3.0	3.0	1.0	1.0	18	10	14	15.0
chr1:71076694-71076893	ENSG00000132485.14|ZRANB2|-	chr1:71076570-71076570|+|ENSG00000132485.14|ZRANB2|124	6.0	2.0	5.0	5.0	1	4	3	0.0
chr1:189437275-189437589	.|.	chr1:189437807-189437807|-|.|218	1.0	0.0	0.0	0.0	298	346	312	374.0
chr1:212615110-212615309	ENSG00000162772.17|ATF3|+	chr1:189437807-189437807|-|.|23177303	0.0	0.0	0.0	0.0	15	4	6	2.0
chr1:212620006-212620205	ENSG00000162772.17|ATF3|+	chr1:189437807-189437807|-|.|23182199	0.0	0.0	0.0	0.0	10	5	15	9.0
chr11:65504957-65505156	ENSG00000251562.8|MALAT1|+	.|-1	1.0	0.0	0.0	0.0	20	11	11	7.0
```

These counts can then be normalized by library size (i.e. number of reads aligning to the hybrid genome) and then a simple 1-tail paired t-test can be used to get significance (p-value) at each integration site. 

Findings:

* There are 9 sites with `nearest_verified_site` having a suffix of `|0` denoting an overlap with a lab-verified integration site. 7 of these are bioionformatically predicted with **High** confidence, 1 with **Medium** confidence and 1 with **Low** confidence. 

* 20 Sites out of 54 *integration windows* have evidence of OST at least in 3 samples. 8 out of the 9 sites from above have evidence of OST.

  